### PR TITLE
[KMSv2] feat: uses decoding buffer for encrypted object

### LINKED ISSUE #118835: kmsv2: use gRPC decoding buffer for encrypted object
This issue refers to the `Marshal` and `Unmarshal` code:

https://github.com/kubernetes/kubernetes/blob/c2fb8057b2b9ad0b278c8f7412a3ec50787dd7e4/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go#L251-L257

We will likely need a struct the encapsulates `proto.NewBuffer` and `EncryptedObject` re-use, probably with one or more `sync.Pool`s.

The benchmarks from #118828 should be used to validate the changes.

### PR DESCRIPTION
**Pull request by nilekhc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR implements sync pools to encapsulate `proto.NewBuffer` and `EncryptedObject` for re-use

#### Which issue(s) this PR fixes:

Fixes #118835

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig auth
/triage accepted
/cc @aramase 

FYI @enj @ritazh

### CODE REVIEW

**Comment by nilekhc** [comment]
@enj How can I run the benchmarks again?

**Comment by enj** [comment]
Looks like this broke something.

```
       <*errors.StatusError | 0xc0030da280>: 
      Internal error occurred: cipher: message authentication failed
      {
          ErrStatus: 
              code: 500
              details:
                causes:
                - message: 'cipher: message authentication failed'
              message: 'Internal error occurred: cipher: message authentication failed'
              metadata: {}
              reason: InternalError
              status: Failure,
      } 
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 1]
```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
 	if err := validateEncryptedObject(request); err != nil {
 		return nil, err
 	}
-	return proto.Marshal(request)
+
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+
+	if err := buffer.Marshal(request); err != nil {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:274]: Add a unit test to catch this bug.

```suggestion

	buffer.Reset()
	if err := buffer.Marshal(request); err != nil {
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go [hunk 2]
```diff
@@ -484,7 +485,20 @@ func TestTransformToStorageError(t *testing.T) {
 }
 
 func TestEncodeDecode(t *testing.T) {
-	transformer := &envelopeTransformer{}
+	transformer := &envelopeTransformer{
+		encryptedObjectDecoder: &encryptedObjectDecoder{
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:489]: Create a `newEncryptedObjectDecoder` func to avoid this duplication.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 3]
```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
-	if err := proto.Unmarshal(originalData, o); err != nil {
-		return nil, err
-	}
-	// validate the EncryptedObject
-	if err := validateEncryptedObject(o); err != nil {
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+	buffer.SetBuf(originalData)
+
+	o := t.encryptedObjectDecoder.objectPool.Get().(*kmstypes.EncryptedObject)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:287]: You need to reset this object.  Add a test for this bug as well.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 4]
```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
 func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.EncryptedObject, error) {
-	o := &kmstypes.EncryptedObject{}
-	if err := proto.Unmarshal(originalData, o); err != nil {
-		return nil, err
-	}
-	// validate the EncryptedObject
-	if err := validateEncryptedObject(o); err != nil {
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+	buffer.SetBuf(originalData)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:285]: I think we can remove the use of `bufferPool` in this path because the `SetBuf` basically makes the buffer not re-use any pre-allocated state (keep using `objectPool`).

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 5]
```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
 }
 
 // doDecode decodes the byte array to an EncryptedObject.
 func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.EncryptedObject, error) {
-	o := &kmstypes.EncryptedObject{}
-	if err := proto.Unmarshal(originalData, o); err != nil {
-		return nil, err
-	}
-	// validate the EncryptedObject
-	if err := validateEncryptedObject(o); err != nil {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:261]: Undo this change, and add a unit test to confirm that we can't drop these validations and have nothing fail.  cc @aramase

**Review by enj — CHANGES_REQUESTED** [review]

**Comment by enj** [comment]
@nilekhc copy the new benchmarks from #118828 and run them like so (remove the KDF bits as needed):

```
go test ./test/integration/controlplane/transformation -bench BenchmarkKMSv2KDF -benchmem -run lolnot -count=10 > old.txt
```

Example of how to compare with `master`: https://github.com/kubernetes/kubernetes/pull/117059#issuecomment-1552163898

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 6]
```diff
@@ -125,6 +132,18 @@ func newEnvelopeTransformerWithClock(envelopeService kmsservice.Service, provide
...
 		cache:           newSimpleCache(clock, cacheTTL),
+		encryptedObjectDecoder: &encryptedObjectDecoder{
+			bufferPool: &sync.Pool{
+				New: func() interface{} {
+					return proto.NewBuffer(nil)
+				},
+			},
+			objectPool: &sync.Pool{
+				New: func() interface{} {
+					return &kmstypes.EncryptedObject{}
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:143]: These objects are a couple bytes. Do we really need to reuse them? The benchmarks for kmsv2 showed that the allocations were dominated everything else in the kube-apiserver. I am skeptical that savings here warrant the additional complexity.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 7]
```diff
@@ -248,17 +267,27 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
-	// validate the EncryptedObject
-	if err := validateEncryptedObject(o); err != nil {
+	buffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(buffer)
+	buffer.SetBuf(originalData)
+
+	o := t.encryptedObjectDecoder.objectPool.Get().(*kmstypes.EncryptedObject)
+	defer t.encryptedObjectDecoder.objectPool.Put(o)
+
+	if err := buffer.Unmarshal(o); err != nil {
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:290]: What do you expect this to do? This does basically the same what we're already doing, with addtional complexity:

https://github.com/kubernetes/kubernetes/blob/4954c7bac4029d2f2e4b305fdba41f81b718aefc/vendor/github.com/gogo/protobuf/proto/decode.go#L396-L402

vs

https://github.com/kubernetes/kubernetes/blob/4954c7bac4029d2f2e4b305fdba41f81b718aefc/vendor/github.com/gogo/protobuf/proto/decode.go#L334-L338

Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:290]: See https://github.com/kubernetes/kubernetes/pull/119167#discussion_r1260181086

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 8]
```diff
@@ -260,22 +286,45 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
 
 // doDecode decodes the byte array to an EncryptedObject.
-func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.EncryptedObject, error) {
-	o := &kmstypes.EncryptedObject{}
-	if err := proto.Unmarshal(originalData, o); err != nil {
+func (t *envelopeTransformer) doDecode(originalData []byte, o *kmstypes.EncryptedObject) (*kmstypes.EncryptedObject, error) {
+	buffer := proto.NewBuffer(originalData)
+
+	resetKMSv2EncryptedObject(o)
+	if err := buffer.Unmarshal(o); err != nil {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:306]: No need for the `buffer` here:

```suggestion
	resetKMSv2EncryptedObject(o)
	if err := proto.Unmarshal(originalData, o); err != nil {
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 9]
```diff
@@ -260,22 +286,45 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
 		return nil, err
 	}
+
+	// validate the EncryptedObject
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:310]:
```suggestion
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 10]
```diff
@@ -260,22 +286,45 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
...
+
+	encodingBuffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(encodingBuffer)
+
+	encodingBuffer.Reset()
+	if err := encodingBuffer.Marshal(request); err != nil {
+		return nil, err
+	}
+
+	return encodingBuffer.Bytes(), nil
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:298]: This `defer` and `encodingBuffer.Bytes()` are not compatible with each other - `Bytes()` just returns the underlying slice which could be re-used before the caller is done with it.  We have a test gap here because nothing is failing.

```suggestion
	encodingBuffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
	defer t.encryptedObjectDecoder.bufferPool.Put(encodingBuffer)

	encodingBuffer.Reset()
	if err := encodingBuffer.Marshal(request); err != nil {
		return nil, err
	}

	return bytes.Clone(encodingBuffer.Bytes()), nil
```

**Review by enj — CHANGES_REQUESTED** [review]

**Comment by nilekhc** [comment]
> @nilekhc copy the new benchmarks from #118828 and run them like so (remove the KDF bits as needed):
> 
> ```
> go test ./test/integration/controlplane/transformation -bench BenchmarkKMSv2KDF -benchmem -run lolnot -count=10 > old.txt
> ```
> 
> Example of how to compare with `master`: [#117059 (comment)](https://github.com/kubernetes/kubernetes/pull/117059#issuecomment-1552163898)


Hi @enj, following is the result of the Benchmark test. The tests were run with `KMSv2KDF=true`. Time taken per operation (sec/op) seem to be improved. But, memory used per operation (B/op), and the number of memory allocations per operation (allocs/op) seem to have been increased. 

Also, I tried to increase the sample size by providing `-benchtime=10s`, but it did not make any difference. Please let me know if I can run the test in any other way.

```shell
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/controlplane/transformation
cpu: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
           │ master.txt  │         usebuffer.txt          │
           │   sec/op    │   sec/op     vs base           │
KMSv2KDF-8   23.46 ± ∞ ¹   19.30 ± ∞ ¹  ~ (p=1.000 n=1) ²
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

           │  master.txt   │          usebuffer.txt           │
           │     B/op      │     B/op       vs base           │
KMSv2KDF-8   18.37Gi ± ∞ ¹   18.84Gi ± ∞ ¹  ~ (p=1.000 n=1) ²
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

           │  master.txt  │          usebuffer.txt          │
           │  allocs/op   │  allocs/op    vs base           │
KMSv2KDF-8   1.122M ± ∞ ¹   1.170M ± ∞ ¹  ~ (p=1.000 n=1) ²
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05
```

**Comment by nilekhc** [comment]
@enj Here are the latest results:

```shell
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/controlplane/transformation
cpu: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
           │ usebuffer_clean.txt │        master_clean.txt               │
           │       sec/op        │   sec/op     vs base                  │
KMSv2KDF-8            17.23 ± 2%   16.94 ± 23%  ~ (p=0.243 n=9+10)

           │ usebuffer_clean.txt │           master_clean.txt            │
           │        B/op         │     B/op      vs base                 │
KMSv2KDF-8          18.84Gi ± 0%   18.37Gi ± 0%  -2.47% (p=0.000 n=9+10)

           │ usebuffer_clean.txt │        master_clean.txt               │
           │      allocs/op      │  allocs/op   vs base                  │
KMSv2KDF-8           1.172M ± 0%   1.174M ± 3%  ~ (p=0.234 n=9+10)

```

**Comment by nilekhc** [comment]
@enj Here are some additional runs,
- with count = 15
```shell
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/controlplane/transformation
cpu: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
           │ usebuffer_15_clean.txt │     master_15_clean.txt             │
           │         sec/op         │   sec/op    vs base                 │
KMSv2KDF-8               17.05 ± 2%   16.95 ± 1%  ~ (p=0.967 n=15)

           │ usebuffer_15_clean.txt │         master_15_clean.txt         │
           │          B/op          │     B/op      vs base               │
KMSv2KDF-8             18.84Gi ± 0%   18.37Gi ± 0%  -2.49% (p=0.000 n=15)

           │ usebuffer_15_clean.txt │        master_15_clean.txt          │
           │       allocs/op        │  allocs/op   vs base                │
KMSv2KDF-8              1.172M ± 0%   1.174M ± 0%  +0.14% (p=0.023 n=15)
```
- with count = 15 and buffer only in decoding

```shell
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/controlplane/transformation
cpu: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
           │ usebuffer_15_clean_decode.txt │        master_15_clean.txt        │
           │            sec/op             │   sec/op    vs base               │
KMSv2KDF-8                      16.70 ± 1%   16.95 ± 1%  +1.53% (p=0.019 n=15)

           │ usebuffer_15_clean_decode.txt │      master_15_clean.txt          │
           │             B/op              │     B/op      vs base             │
KMSv2KDF-8                    18.37Gi ± 0%   18.37Gi ± 0%  ~ (p=0.250 n=15)

           │ usebuffer_15_clean_decode.txt │        master_15_clean.txt        │
           │           allocs/op           │  allocs/op   vs base              │
KMSv2KDF-8                     1.171M ± 0%   1.174M ± 0%  +0.25% (p=0.001 n=15)
```

**Comment by nilekhc** [comment]
@enj another run with count =15 and using buffer in only encoding

```shell
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/controlplane/transformation
cpu: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
           │ usebuffer_15_clean_encode.txt │     master_15_clean.txt             │
           │            sec/op             │   sec/op    vs base                 │
KMSv2KDF-8                      16.80 ± 1%   16.95 ± 1%  ~ (p=0.126 n=15)

           │ usebuffer_15_clean_encode.txt │         master_15_clean.txt         │
           │             B/op              │     B/op      vs base               │
KMSv2KDF-8                    18.84Gi ± 0%   18.37Gi ± 0%  -2.48% (p=0.000 n=15)

           │ usebuffer_15_clean_encode.txt │        master_15_clean.txt          │
           │           allocs/op           │  allocs/op   vs base                │
KMSv2KDF-8                     1.176M ± 0%   1.174M ± 0%  -0.14% (p=0.004 n=15)

```

**Comment by nilekhc** [comment]
We have not seen any significant benefits, so closing this for now. If we see any evidence of performance degradation, we can look into this again.

/close
